### PR TITLE
[SPARK-25837] [Core] Fix potential slowdown in AppStatusListener when cleaning up stages

### PR DESCRIPTION
**Pull request by patrickbrownsync**
## What changes were proposed in this pull request?

* Update `AppStatusListener` `cleanupStages` method to remove tasks for those stages in a single pass instead of 1 for each stage.
* This fixes an issue where the cleanupStages method would get backed up, causing a backup in the executor in ElementTrackingStore, resulting in stages and jobs not getting cleaned up properly.

Tasks seem most susceptible to this as there are a lot of them, however a similar issue could arise in other locations the `KVStore` `view` method is used. A broader fix might involve updates to `KVStoreView` and `InMemoryView` as it appears this interface and implementation can lead to multiple and inefficient traversals of the stored data.

## How was this patch tested?

Using existing tests in AppStatusListenerSuite

This is my original work and I license the work to the project under the project’s open source license.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 1]
```diff
@@ -1105,6 +1095,15 @@ private[spark] class AppStatusListener(
 
       cleanupCachedQuantiles(key)
     }
+
+    // Delete tasks for all stages in one pass, as deleting them for each stage individually is slow
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: Hi @patrickbrownsync , can you explain more about why deleting them in the loop above is slower?  I don't quite understand it.

**Comment by gengliangwang** [comment]
Seems reasonable. Ping @vanzin

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 2]
```diff
@@ -1105,6 +1095,15 @@ private[spark] class AppStatusListener(
...
+    val tasks = kvstore.view(classOf[TaskDataWrapper]).asScala
+    val keys = stages.map(s => (s.info.stageId, s.info.attemptId)).toSet
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: nit: `.map { s => ... }`

**Comment by vanzin** [comment]
ok to test

**Comment by vanzin** [comment]
btw I'm well aware that `InMemoryView` is kinda slow on the read side; improvements there would be welcome. The main issue is that the write path should be as fast as possible, since that can actually cause data loss if you have a burst of new events.

**Comment by patrickbrownsync** [comment]
I would be happy to start thinking about that.

Can you provide more details on what you mean by data loss on write in the InMemoryStore?

I now see that the types being inserted have an annotation around which fields they are allowed to be indexed by, but I don't have a good idea of how you would allow multiple indexes in a collection short of running something like SQLite in memory.

**Review by gengliangwang — APPROVED** [review]
LGTM

**Comment by vanzin** [comment]
Merging to master / 2.4.

**Comment by vanzin** [comment]
(Merged to 2.3 also.)

**Pull request closed without merging**